### PR TITLE
Making it more modular to avoid cyclic dependecy injection

### PR DESCRIPTION
**Pull request by brunodequeiroz**
I was replacing my redux storage to use this module but I was having trouble to avoid cyclic dependency injection, so I though that maybe its better to explicit tell which store the `Provider` should use and pass the react context along to its childrens so the connect function can use it to pass to the Consumer as props. Please let me know what you guys think.

### CODE REVIEW

File: src/Provider.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    const { store } = this.props;
+    if (this.state !== this.value.state) {
+      // If state was changed then recreate `this.value` so it will have a different reference
+      // Explained here: https://reactjs.org/docs/context.html#caveats
+      this.value = { actions: store.actions, state: this.state };
+    }
+
+    return (
+      <store.context.Provider value={this.value}>
+        {React.Children.map(this.props.children, child => React.cloneElement(child, { context: store.context }), {})}
```
Comment (brunodequeiroz) [inline src/Provider.js]: big change

File: src/connect.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import React, { forwardRef } from 'react';
+import Consumer from './Consumer';
+
+const connect: Function = mapStateToProps => WrappedComponent => {
+  const ConnectComponent = forwardRef((props, ref) =>
+    <Consumer context={props.context} mapStateToProps={mapStateToProps}>
```
Comment (brunodequeiroz) [inline src/connect.js]: big change

**Comment by brunodequeiroz** [comment]
Also created a PR with an example using react-router 4 https://github.com/brunodequeiroz/react-waterfall/pull/1

File: README.md [hunk 3]
```diff
@@ -9,31 +9,29 @@ yarn add react-waterfall
 Here is an example of a basic implementation.
 
 ```js
-import { initStore } from 'react-waterfall'
+import { initStore, connect, Provider } from 'react-waterfall';
```
Comment (didierfranc) [inline README.md]: It seems your prettier is not configured to use the project config 

https://github.com/didierfranc/react-waterfall/blob/master/.prettierrc

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-waterfall",
-  "version": "3.0.1",
+  "version": "3.1.0",
```
Comment (didierfranc) [inline package.json]: It will be 4.0.0 because it introduces bc breaking changes

**Comment by didierfranc** [comment]
This is a great proposal, it could be great to add test from @elisherer (https://github.com/didierfranc/react-waterfall/pull/22).

**Comment by elisherer** [comment]
I don't like this idea, because it degenerates the `Consumer` (or `connect`) to just a be a "preventer".
I also didn't understand the cloning and injection of `context` prop into direct children of the `Provider`, this is costly (clone on every state change) and would rarely be the case on a normal application, for example:

```
const App = () => (
<Provider store={store}>
  <Header />
  <Content />
</Provider>
)
const Header = ({ title }) => (
  <div>
     <h1>{title}</h1>
     <MyConnectedComponent />
  </div>
)
```
In this example `MyConnectedComponent` wouldn't have any way to affiliate with the react-waterfall store created above. You would have to manually provide `<MyConnectedComponent context={store.context} />` (And then what's the point if you already have the underlying context).

**Comment by brunodequeiroz** [comment]
Merged the example using `react-router` 4 to this PR

**Comment by didierfranc** [comment]
I close this PR for now, we won't merge it in short term, this is too far from the initial proposal of something thin. Maybe you could publish your own lib based on this PR.

**Pull request closed without merging**